File: phases/phase-6/completed/tools/news.js
/**
 * tools/news.js
 * Phase 5 — 新闻获取工具
 *
 * 主路径：通过 MCP 协议连接 feed-mcp RSS Server
 * 降级路径：直接 fetch RSS XML 并简单解析（若 MCP 不可用）
 */

import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js';

// ─── 预置 RSS 源 ──────────────────────────────────────────────
const RSS_FEEDS = {
  tech:    'https://hnrss.org/newest?count=10',          // Hacker News (英文科技)
  general: 'https://feeds.bbci.co.uk/news/rss.xml',     // BBC News
  world:   'https://feeds.bbci.co.uk/news/world/rss.xml',
  finance: 'https://feeds.bbci.co.uk/news/business/rss.xml',
  default: 'https://hnrss.org/newest?count=10'
};

// MCP 客户端单例（懒初始化）
let mcpClient = null;
let mcpInitFailed = false;

/**
 * 初始化 MCP Client（连接 feed-mcp RSS Server）
 */
async function initMCP() {
  if (mcpInitFailed) return null;
  if (mcpClient) return mcpClient;

  try {
    const transport = new StdioClientTransport({
      command: 'npx',
      args: ['-y', 'feed-mcp@latest'],
      env: { ...process.env }
    });
    mcpClient = new Client({ name: 'aigc-chatbot', version: '1.0.0' }, { capabilities: {} });
    await mcpClient.connect(transport);
    console.log('[News] MCP RSS Server 已连接');
    return mcpClient;
  } catch (e) {
    console.warn('[News] MCP 初始化失败，将使用直接 RSS 解析:', e.message);
    mcpInitFailed = true;
    mcpClient = null;
    return null;
  }
}

/**
 * 获取新闻条目
 * @param {string} topic  主题关键词（tech/general/finance/world）
 * @returns {Promise<{topic:string, items:Array<{title,link,description}>}>}
 */
export async function getNews(topic = 'general') {
  const feedUrl = RSS_FEEDS[topic.toLowerCase()] || RSS_FEEDS.default;

  // 尝试 MCP 路径
  const client = await initMCP();
  if (client) {
    try {
      const result = await client.callTool({
        name: 'get_feed',
        arguments: { url: feedUrl }
      });
      const items = parseMCPFeedResult(result);
      if (items.length > 0) return { topic, source: feedUrl, items };
    } catch (e) {
      console.warn('[News] MCP callTool 失败，降级到直接解析:', e.message);
    }
  }

  // 降级路径：直接 fetch + 解析 RSS XML
  return fetchAndParseRSS(feedUrl, topic);
}

/**
 * 解析 MCP 返回结果（兼容 feed-mcp 的 content 数组格式）
 */
function parseMCPFeedResult(result) {
  try {
    if (Array.isArray(result.content)) {
      const text = result.content.find(c => c.type === 'text')?.text || '';
      return parseRSSXML(text);
    }
    if (typeof result === 'string') return parseRSSXML(result);
    if (Array.isArray(result)) return result.slice(0, 8).map(formatItem);
  } catch {}
  return [];
}

/**
 * 直接 fetch RSS 并解析
 */
async function fetchAndParseRSS(feedUrl, topic) {
  const response = await fetch(feedUrl, {
    headers: { 'User-Agent': 'aigc-chatbot/1.0', 'Accept': 'application/rss+xml, application/xml, text/xml' },
    signal: AbortSignal.timeout(10000)
  });

  if (!response.ok) throw new Error(`RSS 获取失败 (${response.status}): ${feedUrl}`);

  const xml = await response.text();
  const items = parseRSSXML(xml);
  return { topic, source: feedUrl, items };
}

/**
 * 简单 RSS XML 解析（不依赖外部库）
 */
function parseRSSXML(xml) {
  const items = [];
  const itemMatches = [...xml.matchAll(/<item[^>]*>([\s\S]*?)<\/item>/gi)];

  for (const match of itemMatches.slice(0, 8)) {
    const itemXml = match[1];

    const title = extractCDATA(itemXml, 'title');
    const link  = extractTag(itemXml, 'link') || extractTag(itemXml, 'guid');
    const desc  = extractCDATA(itemXml, 'description');
    const pubDate = extractTag(itemXml, 'pubDate');

    if (title) {
      items.push({
        title:       decodeHtmlEntities(title.trim()),
        link:        link?.trim() || '',
        description: decodeHtmlEntities(stripHtmlTags(desc || '')).slice(0, 150),
        pubDate:     pubDate?.trim() || ''
      });
    }
  }

  return items;
}

function extractCDATA(xml, tag) {
  const re = new RegExp(`<${tag}[^>]*>(?:<!\\[CDATA\\[([\\s\\S]*?)\\]\\]>|([\\s\\S]*?))<\\/${tag}>`, 'i');
  const m = xml.match(re);
  return m ? (m[1] ?? m[2] ?? '') : '';
}

function extractTag(xml, tag) {
  const re = new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`, 'i');
  const m = xml.match(re);
  return m ? m[1] : '';
}

function stripHtmlTags(html) {
  return html.replace(/<[^>]+>/g, '').trim();
}

function decodeHtmlEntities(str) {
  return str
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g,  '<')
    .replace(/&gt;/g,  '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g,  "'")
    .replace(/&nbsp;/g, ' ');
}

function formatItem(item) {
  return {
    title:       item.title || item.name || '(无标题)',
    link:        item.url || item.link || '',
    description: item.summary || item.description || ''
  };
}

File: python/phases/phase-5/completed/tools/news.py
"""
新闻工具 — MCP 主路径 + RSS 直接抓取降级
"""
import asyncio
import re
import requests

# RSS 源（国内可访问）
RSS_FEEDS: dict[str, str] = {
    "tech":    "https://hnrss.org/frontpage",
    "general": "https://www.oschina.net/news/rss",
    "world":   "https://rss.dw.com/xml/rss-zh-all",
    "finance": "https://36kr.com/feed",
}
MAX_ITEMS = 10

# 一旦 MCP 失败，后续请求不再重试
_mcp_failed = False


# ── 内部辅助 ─────────────────────────────────────────────

def _extract_tag(xml: str, tag: str) -> str:
    m = re.search(
        rf"<{tag}[^>]*>(?:<!\[CDATA\[([\s\S]*?)\]\]>|([^<]*))",
        xml, re.I,
    )
    if not m:
        return ""
    return (m.group(1) or m.group(2) or "").strip()


def _parse_rss_xml(xml: str) -> list[dict]:
    items: list[dict] = []
    for m in re.finditer(r"<item[^>]*>([\s\S]*?)</item>", xml, re.I):
        item_xml = m.group(1)
        title = _extract_tag(item_xml, "title")
        link  = _extract_tag(item_xml, "link") or _extract_tag(item_xml, "guid")
        pub   = _extract_tag(item_xml, "pubDate")
        desc  = re.sub(r"<[^>]+>", "", _extract_tag(item_xml, "description"))[:200]
        if title:
            items.append({
                "title":   title,
                "link":    link,
                "pubDate": pub,
                "summary": desc,
            })
        if len(items) >= MAX_ITEMS:
            break
    return items


def _rss_fallback(feed_url: str) -> list[dict]:
    resp = requests.get(
        feed_url, timeout=10, headers={"User-Agent": "aigc-chatbot/2.0"}
    )
    resp.raise_for_status()
    return _parse_rss_xml(resp.text)


async def _mcp_fetch(feed_url: str) -> list[dict]:
    from mcp import ClientSession, StdioServerParameters
    from mcp.client.stdio import stdio_client

    server_params = StdioServerParameters(
        command="npx", args=["-y", "feed-mcp"]
    )
    async with stdio_client(server_params) as (read, write):
        async with ClientSession(read, write) as session:
            await session.initialize()
            result = await session.call_tool("get_feed", {"url": feed_url})
            text = ""
            if hasattr(result, "content"):
                for block in result.content:
                    if hasattr(block, "text"):
                        text += block.text
            return _parse_rss_xml(text)


# ── 公开接口 ──────────────────────────────────────────────

def get_news(topic: str = "general") -> dict:
    """
    获取最新新闻资讯。

    Args:
        topic: "tech" | "general" | "world" | "finance"（默认 general）

    Returns:
        {"topic": str, "source": str, "items": list[{title, link, pubDate, summary}]}
    """
    global _mcp_failed
    feed_url = RSS_FEEDS.get(topic.lower(), RSS_FEEDS["general"])

    if not _mcp_failed:
        try:
            # nest_asyncio 解决 Streamlit/Windows 事件循环嵌套问题
            try:
                import nest_asyncio
                nest_asyncio.apply()
            except ImportError:
                pass
            items = asyncio.run(_mcp_fetch(feed_url))
            if items:
                return {"topic": topic, "source": feed_url, "items": items}
        except Exception as e:
            print(f"[News] MCP 失败，降级到直接 RSS：{e}")
            _mcp_failed = True

    items = _rss_fallback(feed_url)
    return {"topic": topic, "source": feed_url, "items": items}
